### ISSUE TITLE
Containerised docker-compose caches the yaml file

### ISSUE THREAD

**Issue by giftig** [docker/compose#3424]
It looks like when using the containerised docker-compose with its bash script wrapper (I need to use this as I'm on OSX Mountain Lion and docker-compose doesn't work on my platform), there's some strange caching going on which breaks my yaml file if I try to re-run after changing it. I get errors which often refer to things I've removed from my docker-compose.yml file, and frequently get this error:

```
ERROR: yaml.reader.ReaderError: unacceptable character #x0000: special characters are not allowed
  in "./docker-compose.yaml", position 607
```

The work-around I've been using for this is to move the file from .yml to .yaml or vice verca; each time I do that and run docker-compose it seems to invalidate whatever was caching it and re-read it, but it's a real pain to have to do this repeatedly when I'm debugging a compose config, especially as I also get other, confusing errors referring to previous states of the file.

Strangely, this cache seems to persist even when I remove the docker-compose container and its image and have it rebuilt from scratch.

**Comment by dnephin** [comment]
I think the issue might be with the virtualbox shared folder used by docker-machine.  We've heard reports of this when using volumes.

**Comment by LennardWesterveld** [comment]
Same problem here changing the file name wil help.. weird problem

**Comment by LennardWesterveld** [comment]
I have the problem when i use the run.sh file because that uses docker-compose docker container.
When i just download the Darwin file form docker-compose repo every thing is working fine for me.

**Comment by evitolins** [comment]
I am experiencing this same issue on 10.8.5 (Mountain Lion)... using the docker-compose container.  Glad I found this.  @giftig idea of switching file ext's definitely helped, but obviously not idea.

**Comment by giftig** [comment]
I suspect the need for a containerised docker-compose is limited to few enough platforms that this issue isn't going to be ranked very highly priority-wise, but hopefully my workaround will help others who have run into this problem, then.

**Comment by codingisacopingstrategy** [comment]
I’m on OS X 10.11, using the Docker Toolbox, and I do have this error when I use Pycharm’s docker support. No problems when running `docker-compose up` from the command line. Renaming the file doesn’t work since it drives Pycharm crazy—it references the Python interpreter by the the docker-compose file. Who knows another way to manually invalidate the cache?  Thanks!

**Comment by benmag** [comment]
This is pretty specific to my situation but it may be of use to someone else so I thought I'd share. 

I was using PHPs `yaml_parse_file` and getting the same error. The workaround I found was to read the file contents with `file_get_contents` and then parse it with `yaml_parse`:

```php
$yaml = yaml_parse(file_get_contents("hello-world.yml"));
```

**Comment by graysonwright** [comment]
I've been having this same problem (OS X, VirtualBox, `docker-machine`) – whenever I need to reset the cache, I run

```bash
docker-machine restart
```

**Comment by graysonwright** [comment]
The same issue is being discussed here:
https://youtrack.jetbrains.com/issue/PY-19944

Roberto Aguilar takes a pretty deep look at the issue in that thread.

It looks like their team has found and included a fix in the `pycharm/docker/compose:1.6.2` image, but I haven't had time to see what that fix entails.

**Comment by graysonwright** [comment]
I just set up a fallback for my machine, which checks to see if a native `docker-compose` has been installed through Homebrew. If it exists, it uses that executable. If it doesn't exist, it falls back to the containerized version that I have on my `PATH`.

In Fish Shell syntax, it is:
```
function dc
  if test -e /usr/local/bin/docker-compose
    # Use the native docker-compose installed by homebrew if it's available.
    # This helps us avoid a caching issue
    # that exists in the containerized version of `docker-compose`
    # See https://github.com/docker/compose/issues/3424
    /usr/local/bin/docker-compose $argv
  else
    docker-compose $argv
  end
end
```
